nca/CoreDS/TcpLikeProperties.py

```diff
@@ -317,6 +366,271 @@ def _get_first_item_str(self):
+                                              paths=paths_dfa, hosts=hosts_dfa,
+                                              icmp_type=icmp_type, icmp_code=icmp_code,
+                                              src_peers=base_peer_set.get_peer_interval_of(peer_in_set),
+                                              dst_peers=dst_peers_interval, base_peer_set=base_peer_set)
+                else:
+                    props = TcpLikeProperties(source_ports=src_ports, dest_ports=ports,
+                                              protocols=protocols if protocols else tcp_protocol, methods=methods,
+                                              paths=paths_dfa, hosts=hosts_dfa,
+                                              icmp_type=icmp_type, icmp_code=icmp_code,
+                                              src_peers=src_peers_interval,
+                                              dst_peers=base_peer_set.get_peer_interval_of(peer_in_set),
+                                              base_peer_set=base_peer_set)
+                if tcp_properties:
+                    tcp_properties |= props
+                else:
+                    tcp_properties = props
+        return tcp_properties
+
+    @staticmethod
+    def make_tcp_like_properties_from_dict(peer_container, cube_dict):
+        """
+        Create TcpLikeProperties from the given cube
+        :param PeerContainer peer_container: the set of all peers
+        :param dict cube_dict: the given cube represented as a dictionary
+        :return: TcpLikeProperties
+        """
+        cube_dict_copy = cube_dict.copy()
+        src_ports = cube_dict_copy.pop("src_ports", PortSet(True))
+        dst_ports = cube_dict_copy.pop("dst_ports", PortSet(True))
+        protocols = cube_dict_copy.pop("protocols", None)
+        src_peers = cube_dict_copy.pop("src_peers", None)
+        dst_peers = cube_dict_copy.pop("dst_peers", None)
+        paths_dfa = cube_dict_copy.pop("paths", None)
+        hosts_dfa = cube_dict_copy.pop("hosts", None)
+        methods = cube_dict_copy.pop("methods", None)
+        icmp_type = cube_dict_copy.pop("icmp_type", None)
+        icmp_code = cube_dict_copy.pop("icmp_code", None)
+        assert not cube_dict_copy
+        return TcpLikeProperties.make_tcp_like_properties(peer_container, src_ports=src_ports, dst_ports=dst_ports,
+                                                          protocols=protocols, src_peers=src_peers, dst_peers=dst_peers,
+                                                          paths_dfa=paths_dfa, hosts_dfa=hosts_dfa, methods=methods,
+                                                          icmp_type=icmp_type, icmp_code=icmp_code)
+
+    @staticmethod
+    def make_empty_properties(peer_container=None):
+        return TcpLikeProperties(base_peer_set=peer_container.peer_set.copy() if peer_container else None, create_empty=True)
```

Comment:
It's a copy of PeerSet, not PeerContainer. It's not efficient to copy it inside the init, because make_tcp_like_properties() calls to init multiple times.